ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java

```diff
@@ -80,41 +87,49 @@ public void run() {
             timer.schedule(refreshIngestionResourceValuesTask, 0, REFRESH_INGESTION_RESOURCES_PERIOD);
 
         } catch (Exception e) {
-            log.error(String.format("Error in initializing ResourceManager: %s.", e.getMessage()), e);
+            log.error("Error in initializing ResourceManager.", e);
             throw e;
         }
     }
 
-    public void clean() {
-        ingestionResources.clear();
+    String getIngestionResource(ResourceTypes resourceType) throws Exception {
+        if (!ingestionResources.containsKey(resourceType)) {
+            // When the value is not available, we need to get the tokens from Kusto (refresh):
+            refreshIngestionResources();
+            try{
+                // If the write lock is locked, then the read will wait here.
+                // In other words if the refresh is running yet, then wait until it ends:
+                ingestionResourcesLock.readLock().lock();
+                if (!ingestionResources.containsKey(resourceType)) {
+                    throw new Exception("Unable to get ingestion resources for this type: " + resourceType.getName());
+                }
+            } finally {
+                ingestionResourcesLock.readLock().unlock();
+            }
+        }
+        return ingestionResources.get(resourceType).nextValue();
     }
 
-    public String getKustoIdentityToken() throws Exception {
+    String getKustoIdentityToken() throws Exception {
         if (identityToken == null) {
             refreshIngestionAuthToken();
-            if (identityToken == null) {
-                throw new Exception("Unable to get Identity token");
+            try{
+                authTokenLock.readLock().lock();
+                if (identityToken == null) {
+                    throw new Exception("Unable to get Identity token");
+                }
+            } finally{
+                authTokenLock.readLock().unlock();
             }
         }
         return identityToken;
     }
 
-    public String getIngestionResource(ResourceTypes resourceType) throws Exception {
-        if (!ingestionResources.containsKey(resourceType)) {
-            refreshIngestionResources();
-            if (!ingestionResources.containsKey(resourceType)) {
-                throw new Exception("Unable to get ingestion resources for this type: " + resourceType.getName());
-            }
-        }
-
-        return ingestionResources.get(resourceType).nextValue();
-    }
-
     int getSize(ResourceTypes resourceType){
         return ingestionResources.containsKey(resourceType) ? ingestionResources.get(resourceType).getSize() : 0;
     }
 
-    private void addValue(String key, String value) {
+    private void addIngestionResource(HashMap<ResourceTypes, IngestionResource> ingestionResources, String key, String value) {
```

Comment:
Isn't the ingestionResources a class member? why do we need to pass it?

Reply:
Yes we have ingestionResource as a class member, but here we pass a new instance of the map to this method.
Only after we finish retrieving/parsing the values from the JSON we received from Kusto, we replace the values in the class member map.